### PR TITLE
Improve NPE error messages for missing data

### PR DESCRIPTION
**Pull request by david-waltermire**
# Committer Notes

In the Metaschema data binding code, missing expected data will generate a NullPointerException. This PR improves these errors to provide more information about the cause of the null value.

### All Submissions:

- [ ] Have you followed the guidelines in our [Contributing](https://github.com/usnistgov/metaschema/blob/master/CONTRIBUTING.md) document?
- [ ] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/usnistgov/metaschema/pulls) for the same update/change?
- [ ] Have you squashed any non-relevant commits and commit messages? \[[instructions](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History)\]
- [ ] Do all automated CI/CD checks pass?

### Changes to Core Features:

- [ ] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your core changes, as applicable?
- [ ] Have you included examples of how to use your new feature(s)?
- [ ] Have you updated all website](https://pages.nist.gov/metaschema) and readme documentation affected by the changes you made? Changes to the website can be made in the website/content directory of your branch.

### CODE REVIEW

File: core/src/main/java/gov/nist/secauto/metaschema/core/datatype/markup/flexmark/AbstractMarkupWriter.java [hunk 1]
```diff
@@ -303,7 +303,7 @@ public void writeLink(
     }
 
     if (!node.getTitle().isBlank()) {
-      String title = ObjectUtils.requireNonNull(node.getTitle().unescape());
+      String title = ObjectUtils.notNull(node.getTitle().unescape());
```
Comment (aj-stein-nist) [inline core/src/main/java/gov/nist/secauto/metaschema/core/datatype/markup/flexmark/AbstractMarkupWriter.java:306]: Non-blocking: is there any practical difference between the two?

File: databind/src/main/java/gov/nist/secauto/metaschema/databind/io/json/MetaschemaJsonReader.java [hunk 2]
```diff
@@ -394,7 +394,14 @@ public Object readChoiceGroupItem(Object parentItem, IBoundInstanceModelChoiceGr
...
 
-      JsonNode descriminatorNode = ObjectUtils.requireNonNull(node.get(instance.getJsonDiscriminatorProperty()));
+      String descriminatorProperty = instance.getJsonDiscriminatorProperty();
+      JsonNode descriminatorNode = node.get(descriminatorProperty);
+      if (descriminatorNode == null) {
+        throw new IllegalArgumentException(String.format(
+            "Unable to find descriminator property '%s' for object at '%s'.",
+            descriminatorProperty,
+            JsonUtil.toString(parser)));
+      }
```
Comment (aj-stein-nist) [inline databind/src/main/java/gov/nist/secauto/metaschema/databind/io/json/MetaschemaJsonReader.java]: Minor, non-blocking: its discriminator, right?

```suggestion
      String discriminatorProperty = instance.getJsonDiscriminatorProperty();
      JsonNode discriminatorNode = node.get(discriminatorProperty);
      if (discriminatorNode == null) {
        throw new IllegalArgumentException(String.format(
            "Unable to find discriminator property '%s' for object at '%s'.",
            discriminatorProperty,
            JsonUtil.toString(parser)));
      }
```

File: databind/src/main/java/gov/nist/secauto/metaschema/databind/io/json/MetaschemaJsonReader.java [hunk 3]
```diff
@@ -711,7 +722,12 @@ public List<?> readList() throws IOException {
         IBoundInstanceFlag jsonKey = instance.getItemJsonKey(item);
         assert jsonKey != null;
 
-        String key = ObjectUtils.requireNonNull(jsonKey.getValue(item)).toString();
+        Object keyValue = jsonKey.getValue(item);
+        if (keyValue == null) {
+          throw new IOException(String.format("Null JSON key value for definition '%s'",
```
Comment (aj-stein-nist) [inline databind/src/main/java/gov/nist/secauto/metaschema/databind/io/json/MetaschemaJsonReader.java]: Minor, non-blocking, we are just talking about the value of `json-key` here, right? I thought using its Metaschema syntax name would make it clear (to me, at least).

```suggestion
          throw new IOException(String.format("Null value for json-key for definition '%s'",
```

File: databind/src/main/java/gov/nist/secauto/metaschema/databind/io/json/MetaschemaJsonWriter.java [hunk 4]
```diff
@@ -128,8 +127,13 @@ public void writeItemFieldValue(Object parentItem, IBoundFieldValue fieldValue)
 
       String valueKeyName;
       if (jsonValueKey != null) {
+        Object keyValue = jsonValueKey.getValue(parentItem);
+        if (keyValue == null) {
+          throw new IOException(String.format("Null JSON value key value for definition '%s'",
```
Comment (aj-stein-nist) [inline databind/src/main/java/gov/nist/secauto/metaschema/databind/io/json/MetaschemaJsonWriter.java]: Minor, non-blocking: given the above similar feedback maybe?

```suggestion
          throw new IOException(String.format("Null value for json-value-key for definition '%s'",
```

File: databind/src/main/java/gov/nist/secauto/metaschema/databind/io/json/MetaschemaJsonWriter.java [hunk 5]
```diff
@@ -297,8 +302,14 @@ private <T extends IFeatureComplexItemValueHandler & IBoundInstanceModel> void w
 
       IBoundInstanceFlag jsonKey = handler.getItemJsonKey(parentItem);
       if (jsonKey != null) {
+        Object keyValue = jsonKey.getValue(parentItem);
+        if (keyValue == null) {
+          throw new IOException(String.format("Null JSON key value for definition '%s'",
```
Comment (aj-stein-nist) [inline databind/src/main/java/gov/nist/secauto/metaschema/databind/io/json/MetaschemaJsonWriter.java]: Minor, non-blocking: given the above similar feedback maybe?

```suggestion
          throw new IOException(String.format("Null value for json-key for definition '%s'",
```

File: databind/src/main/java/gov/nist/secauto/metaschema/databind/io/json/MetaschemaJsonWriter.java [hunk 6]
```diff
@@ -323,8 +334,14 @@ private <T extends IFeatureComplexItemValueHandler & IBoundInstanceModelGroupedN
       IBoundInstanceModelChoiceGroup choiceGroup = handler.getParentContainer();
       IBoundInstanceFlag jsonKey = choiceGroup.getItemJsonKey(parentItem);
       if (jsonKey != null) {
+        Object keyValue = jsonKey.getValue(parentItem);
+        if (keyValue == null) {
+          throw new IOException(String.format("Null JSON key value for definition '%s'",
```
Comment (aj-stein-nist) [inline databind/src/main/java/gov/nist/secauto/metaschema/databind/io/json/MetaschemaJsonWriter.java]: Minor, non-blocking: given the above similar feedback maybe?

```suggestion
          throw new IOException(String.format("Null value for json-key for definition '%s'",
```

File: schemagen/src/main/java/gov/nist/secauto/metaschema/schemagen/json/impl/AbstractModelDefinitionJsonSchema.java [hunk 7]
```diff
@@ -71,7 +71,13 @@ protected AbstractModelDefinitionJsonSchema(
 
     // determine the flag instances to generate
     if (jsonKeyFlagName != null) {
-      IFlagInstance jsonKeyFlag = ObjectUtils.requireNonNull(definition.getFlagInstanceByName(jsonKeyFlagName));
+      IFlagInstance jsonKeyFlag = definition.getFlagInstanceByName(jsonKeyFlagName);
+      if (jsonKeyFlag == null) {
+        throw new IllegalArgumentException(
+            String.format("The referenced JSON key flag '%s' does not exist on definition '%s'.",
```
Comment (aj-stein-nist) [inline schemagen/src/main/java/gov/nist/secauto/metaschema/schemagen/json/impl/AbstractModelDefinitionJsonSchema.java]: Minor, non-blocking: given the above similar feedback maybe?

```suggestion
            String.format("The referenced json-key flag-name '%s' does not exist on definition '%s'.",
```

**Review by aj-stein-nist — DISMISSED** [review]
I will stand to benefit more from these than anyone else, so I definitely like what I see. I have minor style suggestions that resonate with me, but are not blocking or significant.

File: core/src/main/java/gov/nist/secauto/metaschema/core/datatype/markup/flexmark/AbstractMarkupWriter.java (hunk 1 above)
Comment (david-waltermire) [reply core/src/main/java/gov/nist/secauto/metaschema/core/datatype/markup/flexmark/AbstractMarkupWriter.java:306]: `notNull` is an assert, while requireNonNull is actually checking for nullness. You can look at the implementations to see the difference.

File: databind/src/main/java/gov/nist/secauto/metaschema/databind/io/json/MetaschemaJsonReader.java (hunk 2 above)
Comment (david-waltermire) [reply databind/src/main/java/gov/nist/secauto/metaschema/databind/io/json/MetaschemaJsonReader.java]: Sadness... Thanks!

**Review by aj-stein-nist — DISMISSED** [review]

**Review by aj-stein-nist — APPROVED** [review]

**Comment by aj-stein-nist** [comment]
Woops, sorry I missed the one variable! 🤦

**Pull request merged**